AlbumArtist Extension v0.5

rdswift opened the pull request:
# AlbumArtist Extension
This plugin provides ***standardized***, ***credited*** and ***sorted*** artist information for the album artist.  This is useful when your tagging or renaming scripts require both the standardized artist name and the credited artist name, or more detailed information about the album artists.

## Output
The information is provided in the following variables:

* **\_aaeStdAlbumArtists** = The standardized version of album artist(s)

* **\_aaeCredAlbumArtists** = The credited version of album artist(s)

* **\_aaeSortAlbumArtists** = The sorted version of album artist(s)

* **\_aaeAlbumArtistCount** = The number of artists comprising the album artist.

* **\_aaeStdAlbumArtist\_n** = The standardized version of the album artists, where **n** is the number of the artist in the list starting at **0**.  If there are two artists in the AlbumArtist tag, then they will be available in the **\_aaeStdAlbumArtist\_0** and **\_aaeStdAlbumArtist\_1** variables. 

* **\_aaeCredAlbumArtist\_n** = The credited version of the album artists, where **n** is the number of the artist in the list starting at **0**.  If there are two artists in the AlbumArtist tag, then they will be available in the **\_aaeCredAlbumArtist\_0** and **\_aaeCredAlbumArtist\_1** variables.

* **\_aaeSortAlbumArtists\_n** = The sorted version of the album artists, where **n** is the number of the artist in the list starting at **0**.  If there are two artists in the AlbumArtist tag, then they will be available in the **\_aaeSortAlbumArtist\_0** and **\_aaeSortAlbumArtist\_1** variables.

* **\_aaeJoinPhrase\_n** = The phrases used to join the album artists, where **n** is the number of the phrase in the list starting at **0**.  **NOTE:** This variable will not be provided if there is only one artist in the AlbumArtist tag.  The user should check that **\_aaeAlbumArtistCount** is greater than one before using this variable.

**PLEASE NOTE:** Tagger scripts are required to make use of these hidden variables. Please see the notes on "Usage" below.

## Usage
The following are examples of simple Picard file naming scripts to illustrate the use of the variables provided by this plugin.

### Example 1
File the album in a subdirectory of the standardized name of the first artist listed in the AlbumArtist tag.

    %_aaeStdAlbumArtist_0%/%album%/%title%

### Example 2
File the album in a subdirectory of the standardized names of all artists listed in the AlbumArtist tag.  This is similar to using the AlbumArtist tag with the 'Use standardized artist name' option checked in Picard.

    %_aaeStdAlbumArtists%/%album%/%title%

### Example 3
File the album in a subdirectory of the credited name of the first artist listed in the AlbumArtist tag.

    %_aaeCredAlbumArtist_0%/%album%/%title%

### Example 4
File the album in a subdirectory of the credited names of all artists listed in the AlbumArtist tag.  This is similar to using the AlbumArtist tag with the 'Use standardized artist name' option unchecked in Picard.

    %_aaeCredAlbumArtists%/%album%/%title%

### Example 5
File the album in a subdirectory of the sort name of the first artist listed in the AlbumArtist tag.

    %_aaeSortAlbumArtist_0%/%album%/%title%


### Example 6
File the album in a subdirectory of the sort names of all artists listed in the AlbumArtist tag.  This is similar to using the AlbumArtistSort tag.

    %_aaeStdAlbumArtists%/%album%/%title%

### Example 7
File the album in a custom subdirectory if there are multiple album artists.

    %_aaeCredAlbumArtist_0%$if($gt(%_aae_AlbumArtistCount,0), (et al.))/%album%/%title%

### Example 8
File the album in a subdirectory of the first character of the sort name of the first artist listed in the AlbumArtist tag, filing by the standardized name of the first album artist.

    $upper($firstalphachar(%_aaeSortAlbumArtist_0%,nonalpha="#"))/%_aaeStdAlbumArtist_0%/%album%/%title%

## Technical Notes
This plugin does not require any additional calls to the MusicBrainz servers.  It uses the information already provided with the original call, including items that are subsequently discarded by Picard.

rdswift committed:
AlbumArtist Extension v0.3

This plugin provides standardized, credited and sorted artist
information for the album artist.  This is useful when your tagging
or renaming scripts require both the standardized artist name and
the credited artist name, or more detailed information about the
album artists.

rdswift committed:
Corrected issue found during code review.

Sophist-UK commented:
IMO the results of this script i.e. the metadata variables are WAY too complicated.

I would personally prefer to see only complete replacements for original strings and "primary" artists (i.e. the first artist in the string).

rdswift committed:
Version 0.4 update

Simplified outputs as requested during review.  Now provides information
for the complete album artist chain and the first (primary) album artist
only.

rdswift committed:
Remove unused variable noted during code review.

rdswift commented:
@Sophist-UK

I've simplified it as you suggested.  The outputs are now:

* **\_aaeStdAlbumArtists** = The standardized version of the album artists

* **\_aaeCredAlbumArtists** = The credited version of the album artists

* **\_aaeSortAlbumArtists** = The sorted version of the album artists

* **\_aaeStdPrimaryAlbumArtist** = The standardized version of the first (primary) album artist.

* **\_aaeCredPrimaryAlbumArtist** = The credited version of the first (primary) album artist.

* **\_aaeSortPrimaryAlbumArtist** = The sorted version of the first (primary) album artist.

* **\_aaeAlbumArtistCount** = The number of artists comprising the album artist.

Thanks for your review and suggestions.  I agree that it's a **lot** cleaner now, and the removed items would likely never be used anyway so won't be missed.

Sophist-UK commented:
1. Is _aaeCredAlbumArtists equal to AlbumArtists?
2. Can you mark this to run after other plugins that change AlbumArtist / AlbumArtists / AlbumArtistSort / ~albumartists_sort?

rdswift commented:
To answer your questions...

1. It depends on whether or not the '***Use standardized artist name***' option is checked in Picard.  If it's checked, then _aaeStdAlbumArtists = AlbumArtists.  If it's not checked, then _aaeCredAlbumArtists = AlbumArtists.  I included both (even though one of the will always be a duplicate of AlbumArtists) so that the information can be retrieved consistently regardless of whether or not the '*Use standardized artist name*' option is checked.

2. I can (and will) change the call to set ***priority=PluginPriority.LOW*** so that it runs later in the queue, but it doesn't currently read or modify any of the existing variables and tags.  For the reasoning above, it can't rely on the contents of _albumartists because that could be either standardized or credited names depending on the setting in Picard.  I suppose that I could read the status of that setting and use the _albumartists values when creating the appropriate (i.e.: standardized or credited) variables.  I could also certainly use the sorted values from _albumartists_sort rather than the values from the XML results passed to the plugin from Picard,  Is this what you're suggesting?

Sophist-UK commented:
Sounds fine.

rdswift committed:
Update to v0.05

Use the _albumartists values when creating either the _aaeStd* or the
_aaeCred* variables depending on the value of the 'Use standardized
artist name' setting.

Use the _albumartists_sort values when creating the _aaeSort* variables.

rdswift commented:
Updated as discussed.  Thanks again.

rdswift committed:
Correct error in plugin description

rdswift commented:
New description after changes from discussion (with usage examples excluded).  The full README is available at [https://github.com/rdswift/picard-plugins/tree/rds_working_1.0/plugins/albumartistextension](https://github.com/rdswift/picard-plugins/tree/rds_working_1.0/plugins/albumartistextension).

# AlbumArtist Extension
This plugin provides ***standardized***, ***credited*** and ***sorted*** artist information for the album artist.  This is useful when your tagging or renaming scripts require both the standardized artist name and the credited artist name, or more detailed information about the album artists.

## Output
The information is provided in the following variables:

* **\_aaeStdAlbumArtists** = The standardized version of the album artists

* **\_aaeCredAlbumArtists** = The credited version of the album artists

* **\_aaeSortAlbumArtists** = The sorted version of the album artists

* **\_aaeStdPrimaryAlbumArtist** = The standardized version of the first (primary) album artist.

* **\_aaeCredPrimaryAlbumArtist** = The credited version of the first (primary) album artist.

* **\_aaeSortPrimaryAlbumArtist** = The sorted version of the first (primary) album artist.

* **\_aaeAlbumArtistCount** = The number of artists comprising the album artist.

## Technical Notes
This plugin does not require any additional calls to the MusicBrainz servers.  It uses the information already provided with the original call, including items that are subsequently discarded by Picard.

If the '***Use standardized artist name***' setting in Picard is checked the values in the \_albumartists variable are used to prepare the \_aaeStd\* variables, otherwise they are used to prepare the \_aaeCred\* variables.

The values in the \_albumartists\_sort variable are used to prepare the \_aaeSort\* variables.

The plugin is registered at a ***priority=LOW*** setting so that it is executed later in the queue, thus allowing other plugins that modify the values in the \_albumartists and \_albumartists\_sort variables to execute first.  This way the most up-to-date values are used by AlbumArtist Extension when creating its output variables.
